src/runtime/float.cpp

```diff
@@ -33,12 +35,37 @@ extern "C" PyObject* PyFloat_FromString(PyObject* v, char** pend) noexcept {
 }
 
 extern "C" double PyFloat_AsDouble(PyObject* o) noexcept {
-    if (o->cls == float_cls)
+    assert(o);
+
+    if (PyFloat_Check(o))
         return static_cast<BoxedFloat*>(o)->d;
-    else if (isSubclass(o->cls, int_cls))
-        return static_cast<BoxedInt*>(o)->n;
-    Py_FatalError("unimplemented");
-    return 0.0;
+
+    // special case: int (avoids all the boxing below, and int_cls->tp_as_number
+    // isn't implemented at the time of this writing)
+    if (PyInt_Check(o))
+        return (double)static_cast<BoxedInt*>(o)->n;
+    // special case: long
+    if (PyLong_Check(o))
```

Comment:
Ah... is `tp_as_number` overrideable? I don't really get how inheritance of
the tp_ fields works in general.

On Fri, May 8, 2015 at 1:43 PM, Kevin Modzelewski <notifications@github.com>
wrote:

> In src/runtime/float.cpp
> <https://github.com/dropbox/pyston/pull/507#discussion_r29974966>:
>
> >          return static_cast<BoxedFloat*>(o)->d;
> > -    else if (isSubclass(o->cls, int_cls))
> > -        return static_cast<BoxedInt*>(o)->n;
> > -    Py_FatalError("unimplemented");
> > -    return 0.0;
> > +
> > +    // special case: int (avoids all the boxing below, and int_cls->tp_as_number
> > +    // isn't implemented at the time of this writing)
> > +    if (PyInt_Check(o))
> > +        return (double)static_cast<BoxedInt*>(o)->n;
> > +    // special case: long
> > +    if (PyLong_Check(o))
>
> do you know if these should be CheckExact?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dropbox/pyston/pull/507/files#r29974966>.
>